### PR TITLE
training julia models

### PR DESCRIPTION
**Pull request by ylxdzsw**
This PR fixes bugs of `back!` in core.jl (the fallback method) and `Chain`, and add the support of `Affine`. With these we can run the training process of some basic linear models like LogisticRegression on our pure Julia implementation. This allow us to test training process and optimizers before we can run them on backends.

### CODE REVIEW

File: src/layers/affine.jl [hunk 1]
```diff
@@ -9,3 +9,16 @@ Affine(in::Integer, out::Integer; init = initn) =
...
+  W.Δx[:] = x' * Δ
+  b.Δx[:] = sum(Δ, 1)
+  Δ * W.x'
+end
+
+function update!(m::Affine, η)
+  update!(m.W, η)
+  update!(m.b, η)
+  m
+end
```
Comment (MikeInnes) [inline src/layers/affine.jl]: would be good to have the newline here

File: src/layers/control.jl [hunk 2]
```diff
@@ -7,9 +7,25 @@ end
...
-back!(s::Chain, Δ) = foldr((m, Δ) -> back!(m, Δ), Δ, s.layers)
 update!(s::Chain, η) = foreach(l -> update!(l, η), s.layers)
 
+function back!(s::Chain, Δ, xs...)
+  crumbs = Tuple[xs]
+  N = length(s.layers)
+
+  for i in 1:N-1
+    xs = s.layers[i](xs...)
+    xs isa Tuple || (xs = (xs, ))
```
Comment (MikeInnes) [inline src/layers/control.jl]: we can assume that all layers in a chain are single input/output. Then it should be easy to do this as `fold` agains as well.

File: src/layers/affine.jl (hunk 1 above)
Comment (ylxdzsw) [inline src/layers/affine.jl]: Yes, usually I use [editorconfig](http://editorconfig.org/) for trimming trailling white spaces and insert new line automatically. It's simple and works on almost all editors and platforms. Would you mind me adding one for Flux?

File: src/layers/control.jl (hunk 2 above)
Comment (ylxdzsw) [inline src/layers/control.jl]: OK, that would make things easier.

File: src/layers/affine.jl (hunk 1 above)
Comment (MikeInnes) [inline src/layers/affine.jl]: Better to just keep editor config locally, I think.

**Comment by MikeInnes** [comment]
Great, can you add a test?

**Comment by MikeInnes** [comment]
Thanks!

File: src/layers/affine.jl (hunk 1 above)
Comment (MikeInnes) [inline src/layers/affine.jl]: By the way, this method is already output by `@net`, but may be incorrect. Could you fix that and remove this version?

**Pull request merged**